sql/datetimetype_test.go

```diff
@@ -0,0 +1,224 @@
+package sql
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/query"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDatetimeCompare(t *testing.T) {
+	tests := []struct {
+		typ Type
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{Date, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			time.Date(2012, 12, 12, 12, 24, 24, 24, time.UTC), 0},
+		{Date, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			"2010-06-03", 1},
+		{Date, "2010-06-03 06:03:11",
+			time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC), -1},
+		{Datetime, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			time.Date(2012, 12, 12, 12, 24, 24, 24, time.UTC), -1},
+		{Datetime, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			"2010-06-03", 1},
+		{Datetime, "2010-06-03 06:03:11",
+			time.Date(2010, 6, 3, 6, 3, 11, 0, time.UTC), 0},
+		{Timestamp, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			time.Date(2012, 12, 12, 12, 24, 24, 24, time.UTC), -1},
+		{Timestamp, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
+			"2010-06-03", 1},
+		{Timestamp, "2010-06-03 06:03:11",
+			time.Date(2010, 6, 3, 6, 3, 11, 0, time.UTC), 0},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%v %v", test.val1, test.val2), func(t *testing.T) {
+			cmp, err := test.typ.Compare(test.val1, test.val2)
+			require.NoError(t, err)
+			assert.Equal(t, test.expectedCmp, cmp)
+		})
+	}
+}
+
+func TestDatetimeCreate(t *testing.T) {
+	tests := []struct {
+		baseType query.Type
+		expectedType datetimeType
+		expectedErr bool
+	}{
+		{sqltypes.Date, datetimeType{sqltypes.Date}, false},
+		{sqltypes.Datetime, datetimeType{sqltypes.Datetime}, false},
+		{sqltypes.Timestamp, datetimeType{sqltypes.Timestamp}, false},
+		{sqltypes.Binary, datetimeType{}, true},
```

Comment:
I would separate this out into two tests, one for compatible and one for incompatible types.